Jadx-gui 0.8.0-b569-20b03aa fails on framework.jar with java classes inside

In https://github.com/skylot/jadx/issues/388#issuecomment-437601188 @skylot says that jadx can process java .jar with .class files inside.
I tried to do this on out/target/common/obj/JAVA_LIBRARIES/framework_intermediates/classes.jar (from LineageOS 14.1 after ROM build), and jadx-gui failed on this file:
```
INFO  - output directory: classes
INFO  - loading ...
INFO  - converting to dex: classes.jar ...
WARN  - dx message: 
trouble writing output:
com.android.dex.DexIndexOverflowException: Too many method references to fit in one dex file: 102035; max is 65536.
You may try using multi-dex. If multi-dex is enabled then the list of classes for the main dex list is too large.
References by package:
    27 android
   304 android.accessibilityservice
   685 android.accounts
   774 android.animation
    28 android.annotation
  6966 android.app
  1034 android.app.admin
   153 android.app.assist
   504 android.app.backup
   190 android.app.job
    78 android.app.trust
   185 android.app.usage
   168 android.appwidget
  2280 android.bluetooth
   232 android.bluetooth.le
  2016 android.content
  2286 android.content.pm
    55 android.content.pm.permission
   792 android.content.res
   535 android.database
   499 android.database.sqlite
   116 android.ddm
   169 android.drm
    23 android.filterfw
   817 android.filterfw.core
    26 android.filterfw.format
    42 android.filterfw.geometry
    46 android.filterfw.io
    83 android.filterpacks.base
   410 android.filterpacks.imageproc
     6 android.filterpacks.numeric
    17 android.filterpacks.performance
    18 android.filterpacks.text
    29 android.filterpacks.ui
    18 android.filterpacks.videoproc
    21 android.filterpacks.videosink
    86 android.filterpacks.videosrc
   223 android.gesture
  1515 android.graphics
  1898 android.graphics.drawable
    41 android.graphics.drawable.shapes
   108 android.graphics.pdf
   709 android.hardware
   358 android.hardware.camera2
    19 android.hardware.camera2.dispatch
   427 android.hardware.camera2.impl
   483 android.hardware.camera2.legacy
    25 android.hardware.camera2.marshal
   208 android.hardware.camera2.marshal.impl
   192 android.hardware.camera2.params
   122 android.hardware.camera2.utils
   252 android.hardware.display
   270 android.hardware.fingerprint
   445 android.hardware.hdmi
   231 android.hardware.input
   678 android.hardware.location
   283 android.hardware.radio
   206 android.hardware.soundtrigger
   281 android.hardware.usb
     1 android.icu.impl
     2 android.icu.lang
    16 android.icu.text
    33 android.icu.util
   452 android.inputmethodservice
  1386 android.location
  4118 android.media
   293 android.media.audiofx
   129 android.media.audiopolicy
   114 android.media.browse
    48 android.media.effect
    39 android.media.effect.effects
   286 android.media.midi
   141 android.media.projection
   791 android.media.session
    38 android.media.soundtrigger
  1346 android.media.tv
   257 android.mtp
  2445 android.net
    57 android.net.http
   182 android.net.metrics
   107 android.net.nsd
  1025 android.net.wifi
   275 android.net.wifi.nan
   250 android.net.wifi.p2p
    96 android.net.wifi.p2p.nsd
     1 android.net.wimax
   521 android.nfc
   121 android.nfc.cardemulation
   175 android.nfc.tech
  1788 android.opengl
  3574 android.os
    92 android.os.health
   520 android.os.storage
    11 android.permissionpresenterservice
   912 android.preference
   801 android.print
     6 android.print.pdf
   195 android.printservice
    63 android.printservice.recommendation
   737 android.provider
  2084 android.renderscript
    35 android.sax
   366 android.security
   124 android.security.keymaster
   633 android.security.keystore
   200 android.security.net.config
   147 android.service.carrier
    47 android.service.chooser
   166 android.service.dreams
    44 android.service.gatekeeper
    14 android.service.gesture
   127 android.service.media
   439 android.service.notification
    42 android.service.persistentdata
   123 android.service.quicksettings
     3 android.service.restrictions
    30 android.service.textservice
    82 android.service.trust
   303 android.service.voice
    49 android.service.vr
   140 android.service.wallpaper
   149 android.speech
   609 android.speech.tts
    58 android.system
  1662 android.telecom
  1270 android.telephony
    48 android.telephony.cdma
    14 android.telephony.gsm
    52 android.test
   966 android.text
   118 android.text.format
   389 android.text.method
   426 android.text.style
    50 android.text.util
   604 android.transition
  1036 android.util
    87 android.util.apk
    83 android.util.jar
  6007 android.view
   610 android.view.accessibility
   304 android.view.animation
   544 android.view.inputmethod
   145 android.view.textservice
  1074 android.webkit
  9486 android.widget
   283 com.android.framework.protobuf.nano
    12 com.android.framework.protobuf.nano.android
    24 com.android.i18n.phonenumbers
     2 com.android.i18n.phonenumbers.geocoding
   150 com.android.ims
   562 com.android.ims.internal
    24 com.android.ims.internal.uce
    88 com.android.ims.internal.uce.common
   134 com.android.ims.internal.uce.options
   272 com.android.ims.internal.uce.presence
    62 com.android.ims.internal.uce.uceservice
    24 com.android.internal
    41 com.android.internal.alsa
     7 com.android.internal.annotations
  1744 com.android.internal.app
   255 com.android.internal.app.procstats
   101 com.android.internal.appwidget
   142 com.android.internal.backup
   112 com.android.internal.content
    23 com.android.internal.database
    26 com.android.internal.graphics.drawable
    13 com.android.internal.hardware
     8 com.android.internal.http
   159 com.android.internal.inputmethod
    78 com.android.internal.location
    35 com.android.internal.logging
     7 com.android.internal.messages
    56 com.android.internal.midi
    61 com.android.internal.net
  1210 com.android.internal.os
   668 com.android.internal.policy
    27 com.android.internal.preference
   236 com.android.internal.statusbar
   534 com.android.internal.telecom
  1266 com.android.internal.telephony
    94 com.android.internal.textservice
    26 com.android.internal.transition
   778 com.android.internal.util
    11 com.android.internal.util.cm
   131 com.android.internal.util.cm.palette
   597 com.android.internal.view
    17 com.android.internal.view.animation
   736 com.android.internal.view.menu
  2551 com.android.internal.widget
    42 com.android.net
    20 com.android.okhttp
     1 com.android.okhttp.internal
    11 com.android.org.bouncycastle.asn1
    17 com.android.org.bouncycastle.asn1.x509
     1 com.android.org.bouncycastle.jce
     1 com.android.org.bouncycastle.jce.provider
     2 com.android.org.bouncycastle.util.encoders
     9 com.android.org.bouncycastle.util.io.pem
     9 com.android.org.bouncycastle.x509
    36 com.android.org.conscrypt
   100 com.android.server
    42 com.android.server.backup
    45 com.android.server.net
    13 com.google.android.collect
   313 com.google.android.gles_jni
   170 com.google.android.util
     1 dalvik.bytecode
    92 dalvik.system
     7 dalvik.system.profiler
     7 default
   266 java.io
   483 java.lang
     1 java.lang.invoke
     5 java.lang.ref
    48 java.lang.reflect
    10 java.math
    91 java.net
   124 java.nio
     8 java.nio.channels
    18 java.nio.charset
    85 java.security
    42 java.security.cert
    10 java.security.interfaces
    12 java.security.spec
    43 java.text
   472 java.util
    62 java.util.concurrent
    18 java.util.concurrent.atomic
    21 java.util.concurrent.locks
     2 java.util.function
    14 java.util.jar
    14 java.util.logging
    18 java.util.regex
    44 java.util.zip
    24 javax.crypto
    13 javax.crypto.spec
    33 javax.microedition.khronos.egl
   267 javax.microedition.khronos.opengles
     1 javax.net
    39 javax.net.ssl
     4 javax.security.auth.x500
     7 junit.framework
     8 libcore.icu
     2 libcore.internal
    28 libcore.io
    10 libcore.net
     2 libcore.net.event
    33 libcore.util
    16 org.apache.harmony.dalvik.ddmc
     2 org.apache.harmony.luni.internal.util
     1 org.apache.harmony.xml
     2 org.apache.http.conn
     5 org.apache.http.conn.scheme
    57 org.apache.http.conn.ssl
    27 org.apache.http.params
     3 org.ccil.cowan.tagsoup
    14 org.codeaurora.camera
    16 org.json
     2 org.kxml2.io
    13 org.xml.sax
     1 org.xml.sax.helpers
    34 org.xmlpull.v1
     2 sun.security.jca
     7 sun.security.pkcs
        at com.android.dx.dex.file.MemberIdsSection.orderItems(MemberIdsSection.java:48)
        at com.android.dx.dex.file.UniformItemSection.prepare0(UniformItemSection.java:72)
        at com.android.dx.dex.file.Section.prepare(Section.java:213)
        at com.android.dx.dex.file.DexFile.toDex0(DexFile.java:631)
        at com.android.dx.dex.file.DexFile.toDex(DexFile.java:271)
        at com.android.dx.command.dexer.Main.writeDex(Main.java:865)
        at com.android.dx.command.dexer.Main.runMonoDex(Main.java:322)
        at com.android.dx.command.dexer.Main.runDx(Main.java:288)
        at jadx.core.utils.files.JavaToDex.convert(JavaToDex.java:39)
        at jadx.core.utils.files.InputFile.loadFromJar(InputFile.java:163)
        at jadx.core.utils.files.InputFile.searchDexFiles(InputFile.java:68)
        at jadx.core.utils.files.InputFile.addFilesFrom(InputFile.java:36)
        at jadx.api.JadxDecompiler.loadFiles(JadxDecompiler.java:123)
        at jadx.api.JadxDecompiler.load(JadxDecompiler.java:87)
        at jadx.gui.JadxWrapper.openFile(JadxWrapper.java:34)
        at jadx.gui.ui.MainWindow.openFile(MainWindow.java:177)
        at jadx.gui.ui.MainWindow.open(MainWindow.java:134)
        at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
        at java.awt.EventQueue.access$500(EventQueue.java:97)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.awt.EventQueue$3.run(EventQueue.java:703)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
        at org.GNOME.Accessibility.AtkWrapper$6.dispatchEvent(AtkWrapper.java:715)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
        at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)


ERROR - Error load file: classes.jar
jadx.core.utils.exceptions.JadxRuntimeException: Error load file: classes.jar
        at jadx.api.JadxDecompiler.loadFiles(JadxDecompiler.java:125)
        at jadx.api.JadxDecompiler.load(JadxDecompiler.java:87)
        at jadx.gui.JadxWrapper.openFile(JadxWrapper.java:34)
        at jadx.gui.ui.MainWindow.openFile(MainWindow.java:177)
        at jadx.gui.ui.MainWindow.open(MainWindow.java:134)
        at java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:311)
        at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
        at java.awt.EventQueue.access$500(EventQueue.java:97)
        at java.awt.EventQueue$3.run(EventQueue.java:709)
        at java.awt.EventQueue$3.run(EventQueue.java:703)
        at java.security.AccessController.doPrivileged(Native Method)
        at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:74)
        at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
        at org.GNOME.Accessibility.AtkWrapper$6.dispatchEvent(AtkWrapper.java:715)
        at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:205)
        at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
        at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
        at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
        at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: jadx.core.utils.exceptions.DecodeException: java class to dex conversion error:
 dx exception: Java to dex conversion error, code: 2
        at jadx.core.utils.files.InputFile.loadFromJar(InputFile.java:169)
        at jadx.core.utils.files.InputFile.searchDexFiles(InputFile.java:68)
        at jadx.core.utils.files.InputFile.addFilesFrom(InputFile.java:36)
        at jadx.api.JadxDecompiler.loadFiles(JadxDecompiler.java:123)
        ... 19 common frames omitted
Caused by: jadx.core.utils.exceptions.JadxException: dx exception: Java to dex conversion error, code: 2
        at jadx.core.utils.files.JavaToDex.convert(JavaToDex.java:46)
        at jadx.core.utils.files.InputFile.loadFromJar(InputFile.java:163)
        ... 22 common frames omitted
Caused by: jadx.core.utils.exceptions.JadxException: Java to dex conversion error, code: 2
        at jadx.core.utils.files.JavaToDex.convert(JavaToDex.java:42)
        ... 23 common frames omitted
```
The file (rename .zip to .jar): [classes.zip](https://github.com/skylot/jadx/files/2568615/classes.zip)

Smaller .jar files (like out/target/common/obj/JAVA_LIBRARIES/services_intermediates/classes.jar) works fine in this scenario.

--- a/jadx-core/src/main/java/jadx/core/utils/files/InputFile.java
+++ b/jadx-core/src/main/java/jadx/core/utils/files/InputFile.java
@@ -1,9 +1,14 @@
 package jadx.core.utils.files;
 
+import static jadx.core.utils.files.FileUtils.isApkFile;
+import static jadx.core.utils.files.FileUtils.isZipDexFile;
+
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
 import java.util.ArrayList;
 import java.util.Enumeration;
 import java.util.List;
@@ -11,19 +16,17 @@ import java.util.jar.JarOutputStream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipFile;
 
-import com.android.dex.Dex;
 import org.apache.commons.io.IOUtils;
 import org.jetbrains.annotations.Nullable;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import com.android.dex.Dex;
+
 import jadx.core.utils.AsmUtils;
 import jadx.core.utils.exceptions.DecodeException;
 import jadx.core.utils.exceptions.JadxException;
 import jadx.core.utils.exceptions.JadxRuntimeException;
-
-import static jadx.core.utils.files.FileUtils.isApkFile;
-import static jadx.core.utils.files.FileUtils.isZipDexFile;
 
 public class InputFile {
 	private static final Logger LOG = LoggerFactory.getLogger(InputFile.class);
@@ -52,7 +55,9 @@ public class InputFile {
 			return;
 		}
 		if (fileName.endsWith(".class")) {
-			addDexFile(loadFromClassFile(file));
+			for (Dex dex : loadFromClassFile(file)) {
+				addDexFile(dex);
+			}
 			return;
 		}
 		if (isApkFile(file) || isZipDexFile(file)) {
@@ -65,7 +70,9 @@ public class InputFile {
 				return;
 			}
 			if (fileName.endsWith(".jar")) {
-				addDexFile(loadFromJar(file));
+				for (Dex dex : loadFromJar(file.toPath())) {
+					addDexFile(dex);
+				}
 				return;
 			}
 			if (fileName.endsWith(".aar")) {
@@ -116,11 +123,11 @@ public class InputFile {
 
 							case ".jar":
 								index++;
-								File jarFile = FileUtils.createTempFile(entryName);
-								try (FileOutputStream fos = new FileOutputStream(jarFile)) {
-									IOUtils.copy(inputStream, fos);
+								Path jarFile = Files.createTempFile(entryName, ".jar");
+								Files.copy(inputStream, jarFile);
+								for (Dex dex : loadFromJar(jarFile)) {
+									addDexFile(entryName, dex);
 								}
-								addDexFile(entryName, loadFromJar(jarFile));
 								break;
 
 							default:
@@ -155,15 +162,19 @@ public class InputFile {
 		}
 	}
 
-	private static Dex loadFromJar(File jarFile) throws DecodeException {
+	private static List<Dex> loadFromJar(Path jar) throws DecodeException {
 		JavaToDex j2d = new JavaToDex();
 		try {
-			LOG.info("converting to dex: {} ...", jarFile.getName());
-			byte[] ba = j2d.convert(jarFile.getAbsolutePath());
-			if (ba.length == 0) {
+			LOG.info("converting to dex: {} ...", jar.getFileName());
+			List<byte[]> byteList = j2d.convert(jar);
+			if (byteList.isEmpty()) {
 				throw new JadxException("Empty dx output");
 			}
-			return new Dex(ba);
+			List<Dex> dexList = new ArrayList<>(byteList.size());
+			for (byte[] b : byteList) {
+				dexList.add(new Dex(b));
+			}
+			return dexList;
 		} catch (Exception e) {
 			throw new DecodeException("java class to dex conversion error:\n " + e.getMessage(), e);
 		} finally {
@@ -173,9 +184,9 @@ public class InputFile {
 		}
 	}
 
-	private static Dex loadFromClassFile(File file) throws IOException, DecodeException {
-		File outFile = FileUtils.createTempFile("cls.jar");
-		try (JarOutputStream jo = new JarOutputStream(new FileOutputStream(outFile))) {
+	private static List<Dex> loadFromClassFile(File file) throws IOException, DecodeException {
+		Path outFile = Files.createTempFile("cls", ".jar");
+		try (JarOutputStream jo = new JarOutputStream(Files.newOutputStream(outFile))) {
 			String clsName = AsmUtils.getNameFromClassFile(file);
 			if (clsName == null || !ZipSecurity.isValidZipEntryName(clsName)) {
 				throw new IOException("Can't read class name from file: " + file);

--- a/jadx-core/src/main/java/jadx/core/utils/files/JavaToDex.java
+++ b/jadx-core/src/main/java/jadx/core/utils/files/JavaToDex.java
@@ -1,6 +1,11 @@
 package jadx.core.utils.files;
 
 import java.io.ByteArrayOutputStream;
+import java.nio.file.DirectoryStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
 
 import com.android.dx.command.dexer.DxContext;
 import com.android.dx.command.dexer.Main;
@@ -13,11 +18,12 @@ public class JavaToDex {
 	private static final String CHARSET_NAME = "UTF-8";
 
 	private static class DxArgs extends Arguments {
-		public DxArgs(DxContext context, String dexFile, String[] input) {
+		public DxArgs(DxContext context, String dexDir, String[] input) {
 			super(context);
-			outName = dexFile;
+			outName = dexDir;
 			fileNames = input;
 			jarOutput = false;
+			multiDex = true;
 
 			optimize = true;
 			localInfo = true;
@@ -31,17 +37,29 @@ public class JavaToDex {
 
 	private String dxErrors;
 
-	public byte[] convert(String javaFile) throws JadxException {
+	public List<byte[]> convert(Path jar) throws JadxException {
 		try (ByteArrayOutputStream out = new ByteArrayOutputStream();
 		     ByteArrayOutputStream errOut = new ByteArrayOutputStream()) {
 			DxContext context = new DxContext(out, errOut);
-			DxArgs args = new DxArgs(context, "-", new String[]{javaFile});
-			int result = (new Main(context)).runDx(args);
+			Path dir = Files.createTempDirectory("jadx");
+			DxArgs args = new DxArgs(
+					context,
+					dir.toAbsolutePath().toString(),
+					new String[]{jar.toAbsolutePath().toString()});
+			int result = new Main(context).runDx(args);
 			dxErrors = errOut.toString(CHARSET_NAME);
 			if (result != 0) {
 				throw new JadxException("Java to dex conversion error, code: " + result);
 			}
-			return out.toByteArray();
+			List<byte[]> list = new ArrayList<>();
+	        try (DirectoryStream<Path> ds = Files.newDirectoryStream(dir)) {
+	            for (Path child : ds) {
+	            	list.add(Files.readAllBytes(child));
+	            	Files.delete(child);
+	            }
+	        }
+	        Files.delete(dir);
+	        return list;
 		} catch (Exception e) {
 			throw new JadxException("dx exception: " + e.getMessage(), e);
 		}

